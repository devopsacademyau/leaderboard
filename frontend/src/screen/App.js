import React, { Component } from "react"
import { H1, Heading } from "../component/Heading"
import Summary from "../component/Summary"
import { Chart } from "../component/Chart"
import requestJson from "../service/Service"

const AppContainer = ({ children }) => (
  <div className="center avenir black-80 mw100 mw7-ns pa3 ph5-ns">
    {children}
  </div>
)

const Loading = () => (
  <AppContainer>
    <H1>Loading</H1>
  </AppContainer>
)

const Err = ({ error }) => {
  console.error(`error in App.js: ${error}`)
  return (
    <AppContainer>
      <H1>Something went wrong</H1>
    </AppContainer>
  )
}

const Footer = () => (
  <div className="bt b--black-20 pv3 mt5 f7 measure">
    Made with{" "}
    <span role="img" aria-label="heart">
      ❤️
    </span>{" "}
    by{" "}
    <a
      href="https://github.com/devopsacademyau/leaderboard/graphs/contributors"
      className="link black-80 underline-hover hover-blue"
      rel="noopener noreferrer"
      target="_blank"
    >
      Open Source Contributors
    </a>
  </div>
)
class App extends Component {
  constructor(props, context) {
    super(props, context)
    this.state = { chartData: null, loading: true, error: null }
  }

  async componentDidMount() {
    let error = null
    const chartData = await requestJson(this.props.url).catch(err => {
      error = err
      return null
    })
    this.setState({ chartData, error, loading: false })
  }

  render() {
    const { loading, error, chartData } = this.state
    if (error) {
      return <Err error={error} />
    }
    if (loading) {
      return <Loading />
    }
    const { charts, config, authors, botComments } = chartData
    const repository = config.repository
    const bot = authors[config.botName] || {}
    bot.comments = botComments
    return (
      <AppContainer>
        <Heading repository={repository} />
        <Summary charts={charts} config={config} bot={bot} />
        {charts.map((chart, i) => {
          const props = { chart, authors, config }
          return <Chart key={i} {...props} />
        })}
        <Footer />
      </AppContainer>
    )
  }
}

export { App, AppContainer, Loading, Footer, Err }
