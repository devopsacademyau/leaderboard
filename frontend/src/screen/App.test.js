import React from "react"
import App from "./App"
import Adapter from "enzyme-adapter-react-16"
import { shallow, configure } from "enzyme"

configure({ adapter: new Adapter() })

test("chartData update for good fetch", async () => {
  const promise = new Promise((resolve, reject) => {
    resolve({ ok: true, status: 200, json: () => responseObjFixture })
  })
  global.fetch = jest.fn().mockImplementationOnce(() => promise)
  console.error = jest.fn()
  const app = shallow(<App url="http://example.com" />)
  expect(app.state("loading")).toBeTruthy()
  await promise
  setImmediate(() => {
    expect(app.state("loading")).toBeFalsy()
    expect(app.state("error")).toBeFalsy()
    expect(app.state("chartData")).toEqual(responseObjFixture)
    expect(console.error).toBeCalled()
    console.error.mockRestore()
  })
})

test("error update for error fetch", async () => {
  const promise = new Promise((resolve, reject) => {
    resolve({ ok: false, status: 500, json: () => null })
  })
  global.fetch = jest.fn().mockImplementationOnce(() => promise)
  const app = shallow(<App url="http://example.com" />)
  expect(app.state("loading")).toBeTruthy()
  await promise
  setImmediate(() => {
    expect(app.state("loading")).toBeFalsy()
    expect(app.state("error")).toBeTruthy()
  })
})

const responseObjFixture = {
  repository: {
    name: "my-repo",
    owner: "my-owner",
    url: "https://github.com/my-owner/my-repo"
  },
  charts: [
    {
      title: "My Chart",
      maxCount: 3,
      totalCount: 6,
      points: [
        { author: "a", count: "1" },
        { author: "b", count: "2" },
        { author: "C", count: "3" }
      ]
    }
  ],
  authors: {
    a: {
      login: "a",
      url: "https://github.com/a",
      avatarURL: "https://avatars3.githubusercontent.com/u/22388715?v=4"
    },
    b: {
      login: "b",
      url: "https://github.com/b",
      avatarURL: "https://avatars0.githubusercontent.com/u/6849798?v=4"
    },
    c: {
      login: "c",
      url: "https://github.com/c",
      avatarURL: "https://avatars2.githubusercontent.com/u/32605850?v=4"
    }
  }
}
