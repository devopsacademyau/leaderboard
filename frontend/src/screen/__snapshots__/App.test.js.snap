// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer matches snapshot 1`] = `
<div
  className="center avenir black-80 mw100 mw7-ns pa3 ph5-ns"
/>
`;

exports[`Err matches snapshot 1`] = `
<div
  className="center avenir black-80 mw100 mw7-ns pa3 ph5-ns"
>
  <h1
    className="f2 fw6 pv1 pt4-ns pb3-ns"
  >
    Something went wrong
  </h1>
</div>
`;

exports[`Footer matches snapshot 1`] = `
<div
  className="bt b--black-20 pv3 mt5 f7 measure"
>
  Made with
   
  <span
    aria-label="heart"
    role="img"
  >
    ❤️
  </span>
   
  by
   
  <a
    className="link black-80 underline-hover hover-blue"
    href="https://github.com/devopsacademyau/leaderboard/graphs/contributors"
    rel="noopener noreferrer"
    target="_blank"
  >
    Open Source Contributors
  </a>
</div>
`;

exports[`Loading matches snapshot 1`] = `
<div
  className="center avenir black-80 mw100 mw7-ns pa3 ph5-ns"
>
  <h1
    className="f2 fw6 pv1 pt4-ns pb3-ns"
  >
    Loading
  </h1>
</div>
`;
